ios/brave-ios/Sources/Brave/Frontend/Browser/New Tab Page/Backgrounds/NewTabPageVideoPlayer.swift

```diff
@@ -0,0 +1,292 @@
+class NewTabPageVideoPlayer {
+  var autoplayFinishedEvent: (() -> Void)?
+  var autoplayStartedEvent: (() -> Void)?
+  var playFinishedEvent: (() -> Void)?
+  var playStartedEvent: (() -> Void)?
+  var played25PercentEvent: (() -> Void)?
+  var playCancelledEvent: (() -> Void)?
+
+  var player: AVPlayer?
+
+  private var isPlayStarted = false
+  private var isLoadFinished = false
+  private var isAutoplayRequestedOnce = false
+  private var isAutoplayFinished = false
+  private var timeObserver: Any?
+  private var playerObserver: NSKeyValueObservation?
+
+  private let kMaxAutoplayDuration = 6.0
+
+  private var stopFrame: Int?
+  private var frameRate: Float?
+
+  init(backgroundVideoPath: URL, videoInitiallyVisible: Bool) {
+    // Do not start autoplay if video is not initially visible
+    if !videoInitiallyVisible {
+      isAutoplayRequestedOnce = true
+      isAutoplayFinished = true
+    }
+
+    stopFrame = parseStopFrameFromFilename(filename: backgroundVideoPath.lastPathComponent)
+    resetPlayer(backgroundVideoPath)
+    loadVideoTrackParams(backgroundVideoPath)
+  }
+
+  deinit {
+    cleanupObservers()
+  }
+
+  func resetPlayer(_ backgroundVideoPath: URL?) {
+    guard let backgroundVideoPath = backgroundVideoPath else {
+      cleanupObservers()
+      player = nil
+      return
+    }
+
+    let item = AVPlayerItem(url: backgroundVideoPath)
+    player = AVPlayer(playerItem: item)
+
+    NotificationCenter.default
+      .addObserver(
+        self,
+        selector: #selector(self.playerDidFinishPlaying),
+        name: .AVPlayerItemDidPlayToEndTime,
+        object: item
+      )
+  }
+
+  func startVideoPlayback() {
+    if !isAutoplayFinished {
+      autoplayFinished()
+    }
+
+    player?.isMuted = false
+    player?.seek(to: CMTime.zero)
+    player?.currentItem?.forwardPlaybackEndTime = CMTime()
+
+    let interval = CMTime(
+      seconds: 0.1,
+      preferredTimescale: CMTimeScale(NSEC_PER_SEC)
+    )
+    // Setup observer to check if 25% of the video is played.
+    timeObserver =
+      player?.addPeriodicTimeObserver(forInterval: interval, queue: nil) {
+        [weak self] time in
+        self?.checkPlayPercentage()
+      }
+
+    player?.play()
+    isPlayStarted = true
+
+    playStartedEvent?()
+  }
+
+  func maybeCancelPlay() {
+    if !isAutoplayFinished && isLoadFinished {
+      autoplayFinished()
+      player?.pause()
+    }
+
+    if isPlayStarted {
+      player?.pause()
+      if let timeObserver = timeObserver {
+        player?.removeTimeObserver(timeObserver)
+      }
+      timeObserver = nil
+
+      playCancelledEvent?()
+    }
+    isPlayStarted = false
+  }
+
+  func togglePlay() -> Bool {
+    let isVideoInProgress = isVideoInProgress()
+
+    if isVideoInProgress {
+      player?.pause()
+    } else {
+      player?.play()
+    }
+
+    return !isVideoInProgress
+  }
+
+  func maybeStartAutoplay() {
+    // Start autoplay only once.
```

Comment:
nit: not sure we need this comment as we could rename `isAutoplayRequestedOnce` to `didAutoplay`